### PR TITLE
[FN] Implementation of the SubtractFeeFromAmount

### PR DESCRIPTION
**Pull request by fenix2222**
Currently `SubtractFeeFromAmount` for recipients is not implemented. This adds the functionality to subtract the fee from recipients amount, similar to:

https://github.com/bitcoin/bitcoin/blob/a689c119076c7b8dc5b4dea4539e4cbf5adfb72f/src/wallet/wallet.cpp

Do not merge yet, need to investigate the best approach with this.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 1]
```diff
@@ -356,18 +356,76 @@ protected virtual void AddRecipients(TransactionBuildContext context)
...
-                throw new NotImplementedException("Substracting the fee from the recipient is not supported yet.");
+            var totalRecipients = context.Recipients.Count(r => r.SubtractFeeFromAmount);
 
-            foreach (Recipient recipient in context.Recipients)
-                context.TransactionBuilder.Send(recipient.ScriptPubKey, recipient.Amount);
+            // If we have any recipients that require a fee to be subtracted from the amount, then
+            // calculate fee and evenly distribute it among all recipients. Any remaining fee should be
+            // subtracted from the first recipient.
+            if (totalRecipients > 0)
+            {
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: At the risk of complicating the code, can this logic not go in to the method AddFee and in that method if falling in the condition `if (context.Recipients.Any(r => r.SubtractFeeFromAmount))` then apply this logic to the recipients already in the builder?

**Comment by dangershony** [comment]
Good change FYI quite link to the area in core where this logic is applied   

https://github.com/bitcoin/bitcoin/blob/a689c119076c7b8dc5b4dea4539e4cbf5adfb72f/src/wallet/wallet.cpp#L3013

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs (hunk 1 above)
Comment (fenix2222) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: @dangershony I thought about doing it, but the issue is that I would have to replace send builder then, as transaction send builder is created with UTXO in AddRecipients

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: No that's even more complicated I agree, I thought there would be the ability to have direct access to the outputs, like how @MithrilMan did when adding the secrets, maybe not.

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 2]
```diff
@@ -356,18 +356,76 @@ protected virtual void AddRecipients(TransactionBuildContext context)
...
+                {
+                    var recipient = context.Recipients[i];
+
+                    if (recipient.SubtractFeeFromAmount)
+                    {
+                        // First receiver pays the remainder not divisible by output count.
+                        long feeToSubtract = i == 0 ? remainingFee + recipientFee : recipientFee;
+                        long remainingAmount = recipient.Amount.Satoshi - feeToSubtract;
+                        if (remainingAmount <= 0)
+                            throw new WalletException("Fee is higher than amount to send.");
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Can we perhaps just add the amounts into the exception message?

```suggestion
                            throw new WalletException(string.Format("Fee {0} is higher than amount {1} to send.",feeToSubtract, recipient.Amount.Satoshi));
```

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 3]
```diff
@@ -356,18 +356,76 @@ protected virtual void AddRecipients(TransactionBuildContext context)
...
+            }
         }
 
         /// <summary>
         /// Use the <see cref="FeeRate"/> from the <see cref="walletFeePolicy"/>.
         /// </summary>
         /// <param name="context">The context associated with the current transaction being built.</param>
         protected void AddFee(TransactionBuildContext context)
+        {
+            if (context.Recipients.Any(r => r.SubtractFeeFromAmount))
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Where is this done in Core's code?

(It looks right, just wanna double check)

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs (hunk 2 above)
Comment (fenix2222) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: :+1:

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs (hunk 3 above)
Comment (fenix2222) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Their code is much more complicated, but it is all inside a `while` loop: 
https://github.com/bitcoin/bitcoin/blob/a689c119076c7b8dc5b4dea4539e4cbf5adfb72f/src/wallet/wallet.cpp#L3015

**Comment by fenix2222** [comment]
@fassadlr and @dangershony  I had to refactor logic to make this work better and added a test. Please review

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 4]
```diff
@@ -57,15 +57,15 @@ public Transaction BuildTransaction(TransactionBuildContext context)
...
-                context.TransactionBuilder.Shuffle();
-
-            const int maxRetries = 3;
+            const int maxRetries = 5;
             int retryCount = 0;
 
             TransactionPolicyError[] errors = null;
             while (retryCount <= maxRetries)
             {
+                if (context.Shuffle)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: What does shuffle do internally? its not just setting a flag?

File: src/NBitcoin/TransactionBuilder.cs [hunk 5]
```diff
@@ -779,6 +779,22 @@ public TransactionBuilder Send(Script scriptPubKey, Money amount)
...
+        {
+            List<Builder> builders = this.CurrentGroup.Builders;
+            for (int i = builders.Count - 1; i >= 0; i--)
+            {
+                // All other builders have different return type.
+                if (builders[i].Method.ReturnType == typeof(Money))
+                {
+                    this.CurrentGroup.Builders.RemoveAt(i);
+                }
+            }
```
Comment (noescape00) [inline src/NBitcoin/TransactionBuilder.cs]: Could be done in 1 line: 

```
this.CurrentGroup.Builders = this.CurrentGroup.Builders.Where(x => x.Method.ReturnType != typeof(Money)).ToList();
```

Comment (noescape00) [reply src/NBitcoin/TransactionBuilder.cs]: Since method is called from just 1 place I guess it's better to replace method call with that one line

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 6]
```diff
@@ -372,18 +381,71 @@ protected virtual void AddRecipients(TransactionBuildContext context)
             if (context.Recipients.Any(a => a.Amount == Money.Zero))
                 throw new WalletException("No amount specified.");
 
-            if (context.Recipients.Any(a => a.SubtractFeeFromAmount))
-                throw new NotImplementedException("Substracting the fee from the recipient is not supported yet.");
+            var totalRecipients = context.Recipients.Count(r => r.SubtractFeeFromAmount);
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: unvar here and below

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs (hunk 4 above)
Comment (fenix2222) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: It is a flag we usually set to true, there is a comment against it: "We shuffle transaction outputs by default as it's better for anonymity.". I have just added this change to make it in line with code hat was already merged into 3.0.6 branch.

File: src/NBitcoin/TransactionBuilder.cs (hunk 5 above)
Comment (fenix2222) [reply src/NBitcoin/TransactionBuilder.cs]: I can't replace it as `CurrentGroup` is inaccessible from outside this class, hence I had to add this helper method.

Comment (fenix2222) [reply src/NBitcoin/TransactionBuilder.cs]: > Could be done in 1 line

Sure, but wouldn't it then involve 2 iterations: 1 for `Where` and one for `ToList` instead of just a simple `for` loop?

https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L30
https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L329

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs (hunk 6 above)
Comment (fenix2222) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Done

**Comment by MithrilMan** [comment]
@fenix2222 
Just wanted to point you out to a method in TransactionBuilder that could simplify the code
TransactionBuilder.SubtractFees

/// <summary>
/// Will subtract fees from the previous TxOut added by the last TransactionBuidler.Send() call
/// </summary>

I would expect this method to deduct the fee if called after every .Send you call to send money to a destination address (if it has to subtract fees of course)
Could be worthy of a check.

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 7]
```diff
@@ -372,18 +381,71 @@ protected virtual void AddRecipients(TransactionBuildContext context)
...
+                        long feeToSubtract = i == 0 ? remainingFee + recipientFee : recipientFee;
+                        long remainingAmount = recipient.Amount.Satoshi - feeToSubtract;
+                        if (remainingAmount <= 0)
+                            throw new WalletException($"Fee {feeToSubtract} is higher than amount {recipient.Amount.Satoshi} to send.");
+
+                        recipient.Amount = new Money(remainingAmount);
+                    }
 
-            foreach (Recipient recipient in context.Recipients)
-                context.TransactionBuilder.Send(recipient.ScriptPubKey, recipient.Amount);
```
Comment (MithrilMan) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: based on my comment, if it's true that TransactionBuilder.SubtractFees does what I expect it to do, it would be enough to call context.TransactionBuilder.SubtractFees after this line

Comment (fenix2222) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: @MithrilMan I checked that code but I don't think it is evenly distributing tx fees across all recipients.

It just subtracts total fee from the send. Can you check that code, it is hard to follow

Comment (MithrilMan) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: you may be right, at a first sight it just subtract it from the last output created before the call, maybe @NicolasDorier could shed light on this, it seems transaction builder doesn't supports deducting fees from multiple outputs

I would expect a way to deduct fee from specific outputs and then be able to pay the eventual remaining fees at the end
like `context.TransactionBuilder.Send(recipient.ScriptPubKey, recipient.Amount, deductFee);` where deductFee is a boolean

Comment (NicolasDorier) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: Check this commit where I tried to rewrite it in a clearer way https://github.com/MetacoSA/NBitcoin/pull/763/commits/83b049d6f80d65910776b4c816040ae235c7d569

The basic idea is: I try to find the builder with the Target is affected by the SubstractFee.
I remove the builder from the list, and create the same one but sending less to the output. If there is not enough money remaining in the output (dust), I delete the builder completely.

Comment (NicolasDorier) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: > like context.TransactionBuilder.Send(recipient.ScriptPubKey, recipient.Amount, deductFee); where deductFee is a boolean

I did not wanted to touch anything that was working before so the syntax is in NBitcoin:

```csharp
builder
.Send(1BTC, alice)
.SubstractFee();
```
The call to `SubstractFee` just keep in memory the last builder above it.

Comment (fenix2222) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: > The basic idea is: I try to find the builder with the Target is affected by the SubstractFee.
I remove the builder from the list and create the same one but sending less to the output. If there is not enough money remaining in the output (dust), I delete the builder completely.

I think this is somewhat similar to what I am doing in this PR, i.e. find a send builder and change it with a new amount. A fee is evenly distributed between recipients who have subtract fee flag set and any remaining fee is subtracted from first recipient as per bitcoin client.
